Documentation/design-docs/hw-intrinsics.md

```diff
@@ -12,47 +12,89 @@ There is a design document for the Arm64 intrinsics: https://github.com/dotnet/c
 
 The reference assemblies for the hardware intrinsics live in corefx, but all of the implementation is in the coreclr repo:
 
-* The C# implementation lives in coreclr/src/shared/System/Runtime/Intrinsics. These are little more than skeleton methods that usually invoke themselves recursively (see Indirect invocation support below).
+* The C# implementation lives in coreclr/src/System.Private.CoreLib/shared/System/Runtime/Intrinsics. These are little more than skeleton methods that are only compiled if needed for indirect invocation.
+
+  * Note that they also exist in the dotnet/corefx repo in src\Common\src\CoreLib\System\Runtime\Intrinsic. These are mirrored between the two repositories.
 
 ## C# Implementation
 
-The hardware intrinsics operate on and produce both primitive types (`int`, `float`, etc.) as well as vector types. The vector types are considered platform-agnostic, though not all platforms define intrinsics on all of these types. They are:
+The hardware intrinsics operate on and produce both primitive types (`int`, `float`, etc.) as well as vector types.
+
+### Platform-agnostic vector types
+
+The vector types supported by one or more target ISAs are supported across platforms, though they extent to which operations on them are available and accelerated is dependent on the target ISA. These are:
 
 * `Vector64<T>` - A 64-bit vector of type `T`. For example, a `Vector64<int>` would hold two 32-bit integers.
+  * Note that `Vector64<T>` intrinsics are supported only on Arm64, and these are not supported for `double`. Support could be added for this, but would require additional handling.
 * `Vector128<T>` - A 128-bit vector of type `T`
 * `Vector256<T>` - A 256-bit vector of type `T`
 
-Note that these are generic types, which distinguishes these from native intrinsic vector types. It also somewhat complicates interop, as the runtime currently doesn't support interop for generic types.
-*** Is there an issue for this?? ***
+Note that these are generic types, which distinguishes these from native intrinsic vector types. It also somewhat complicates interop, as the runtime currently doesn't support interop for generic types. See https://github.com/dotnet/coreclr/issues/1685
 
-The C# declaration of a hardware intrinsic ISA class is marked with the `[Intrinsic]` attribute, and the implementations of the intrinsic methods on that class are recursive. When the VM encounters such a method, it will communicate to the JIT that this is an intrinsic method, and will also pass a `mustExpand` flag to indicate that the JIT must generate code.
+Not all intrinsics defined on these types support all primitive type parameters. When not supported, they are expected to throw `NotSupportedException`. This is generally handled by the C# implementation code.
```

Comment:
I believe all methods exposed by `Vector128<T>` and `Vector128` (for example) will always support all types that are allowed by any ISA (this is currently the 10 primitive types).

The individual ISAs will then take concrete `Vector128<T>` (e.g. `Vector128<float>`) which will work if `Isa.IsSupported` returns `true` (and any requirements around constants are met).

I don't believe, under the current design, we will be exposing anything that a user can't determine support for via a property check and I believe we will always be exploding the generic types such that they won't have to question which types are supported.